src/validation/validation.cpp

```diff
@@ -1979,11 +1984,26 @@ bool ConnectBlockPrevalidations(const CBlock &block,
 {
     int64_t nTimeStart = GetStopwatchMicros();
 
-    // Check it again in case a previous version let a bad block in
-    if (!CheckBlock(block, state, !fJustCheck, !fJustCheck))
+    // If we're not in IBD then check it again in case a previous version let a bad block in
+    //
+    // During IBD setBlocksAlreadyChecked is used rather than the internal block flag block.fChecked because
+    // blocks are typcially stored 1000 blocks ahead and get writting to disk but fChecked is not stored
+    // to disk, so when the block is finally read again for processing it ends up going through CheckBlock() again
+    // unnecessarily. So by storing the hash of the block during IBD just after the first successful CheckBlock()
+    // we can avoid doing CheckBlock() again later in IBD and which saves us a significant amount of time.
+    //
+    // In addition, if we had just restarted the node and continued IBD, started a Reindex, or found a new block
+    // then setBlockAlreadyChecked would be empty and we'd have to run CheckBlock() at least one time, which is
+    // correct behavior.
+    if (!setBlocksAlreadyChecked.count(block.GetHash()))
     {
-        return false;
+        if (!CheckBlock(block, state, !fJustCheck, !fJustCheck))
+        {
+            return false;
+        }
```

Comment:
why not just returning `CheckBlock(block, state, !fJustCheck, !fJustCheck)`?

Reply:
because we only want to return if it's false, but keep going if it's true.

Reply:
You right indeed, brain fart of mine. Sorry for the noise